File: src/main/java/com/dude/dms/backend/service/CrudService.java
package com.dude.dms.backend.service;

import com.dude.dms.backend.data.entity.DataEntity;
import org.springframework.data.jpa.repository.JpaRepository;

import javax.persistence.EntityNotFoundException;
import java.util.List;

public abstract class CrudService<T extends DataEntity> {

    protected abstract JpaRepository<T, Long> getRepository();

    public T save(T entity) {
        return getRepository().saveAndFlush(entity);
    }

    public void delete(T entity) {
        if (entity == null) {
            throw new EntityNotFoundException("entity was null");
        }
        getRepository().delete(entity);
    }

    public long count() {
        return getRepository().count();
    }

    public T load(long id) {
        T entity = getRepository().findById(id).orElse(null);
        if (entity == null) {
            throw new EntityNotFoundException("entity was null");
        }
        return entity;
    }

    public T create(T entity) {
        return getRepository().saveAndFlush(entity);
    }

    public List<T> findAll() {
        return getRepository().findAll();
    }
}

File: src/main/kotlin/com/dude/dms/backend/service/CrudService.kt
package com.dude.dms.backend.service

import com.dude.dms.backend.data.DataEntity
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository

abstract class CrudService<T : DataEntity>(private val repository: JpaRepository<T, Long>) {

    open fun create(entity: T): T = repository.saveAndFlush(entity)

    open fun save(entity: T): T = repository.saveAndFlush(entity)

    open fun delete(entity: T) = repository.delete(entity)

    open fun count() = repository.count()

    fun load(id: Long): T? = repository.findById(id).orElse(null)

    open fun findAll(): List<T> = repository.findAll()

    open fun findAll(pageable: Pageable): Page<T> = repository.findAll(pageable)
}
